[Flow] Extend CollapseDimensionsPass

IanWood1 opened the pull request:
iree-flow-collapse-dimensions is currently very limited. The pass doenst support propagating collapse of dimensions through multiple operations.  This adds support to collapsing dimensions through chains of operations.

Also, this will need to be disabled in the specific case of softmax fusion since there is a backend bug (https://github.com/iree-org/iree/issues/17948).

IanWood1 committed:
Extend CollapseDimensionsPass

iree-flow-collapse-dimensions is currently very limited.  This adds
support to collapsing dimensions through chains of operations.

Signed-off-by: Ian Wood <ianwood2024@u.northwestern.edu>

MaheshRavishankar commented on compiler/src/iree/compiler/Dialect/Flow/Transforms/CollapseDimensions.cpp:
```diff
@@ -186,30 +164,237 @@ static bool isEligibleForCollapse(linalg::GenericOp genericOp) {
+};
+} // namespace
+
+void CollapseInfo::updateReassociation() {
```
I think I wrote this logic sometime, but I dont follow now what it is doing. Some extensive documentation here would really help further down the line.

MaheshRavishankar commented on compiler/src/iree/compiler/Dialect/Flow/Transforms/CollapseDimensions.cpp:
```diff
@@ -186,30 +164,237 @@ static bool isEligibleForCollapse(linalg::GenericOp genericOp) {
+}
+
+FailureOr<CollapseInfo::CollapsableLoopsSet>
+CollapseInfo::getTransformedCollapsableLoops(AffineMap map) const {
```
Here too... write an example of what the transformedloops would be for a given map and collapsableLoops.

MaheshRavishankar commented on compiler/src/iree/compiler/Dialect/Flow/Transforms/CollapseDimensions.cpp:
```diff
@@ -416,44 +571,200 @@ hoistTensorReshapesOutOfDispatchRegion(RewriterBase &rewriter,
+}
+
+// For each producer, use it's consumers to constrain which dimensions it will
+// collapse. `slice` is expected to be topologically sorted (getBackwardSlice
```
THis uses forward slice?

IanWood1 replied:
No, I just use `llvm::reverse` on the slice created by getBackwardSlice so that each ops producer has been updated before before updating said op.

MaheshRavishankar commented on compiler/src/iree/compiler/Dialect/Flow/Transforms/test/collapse_dimensions.mlir:
```diff
@@ -57,3 +58,263 @@ util.func public @unpack_collapse(%arg0: tensor<2x320x128x128xf32>, %arg1: tenso
+//       CHECK:    %[[ELEMENTWISE:.+]] = linalg.generic
+//  CHECK-SAME:      tensor<2x320x16384xf32>
+//       CHECK:    %[[GEN:.+]] = linalg.generic
+//  CHECK-SAME:      tensor<2x320x16384xf32>, tensor<320xf32>, tensor<2x320xf32>, tensor<320xf32>
```
I'd make this check very verbose... essentially I think it is good to check everything that is generated by logic added in the PR (all the new collapsed indexing maps, all the new values, etc.)

MaheshRavishankar reviewed (changes requested):
Quite nice! I want to just understand the logic better, so asking for some more documentation.

IanWood1 committed:
Add more checks to tests and edit isEligibleForCollapse

There is a backend crash when aggressive fusion is enabled and softmax
is fused. This occurs when there is a reduction op that is used in a
broadcast-like way in a subsequent parallel generic. So, collapse is
disabled in that specific case

Signed-off-by: Ian Wood <ianwood2024@u.northwestern.edu>

IanWood1 committed:
add comments with examples

Signed-off-by: Ian Wood <ianwood2024@u.northwestern.edu>

IanWood1 committed:
Add testcase and fix loop map logic

Signed-off-by: Ian Wood <ianwood2024@u.northwestern.edu>

MaheshRavishankar reviewed (approved):
Nice! I think I follow the logic here.

MaheshRavishankar commented on compiler/src/iree/compiler/Dialect/Flow/Transforms/test/collapse_dimensions.mlir:
```diff
@@ -55,5 +56,344 @@ util.func public @unpack_collapse(%arg0: tensor<2x320x128x128xf32>, %arg1: tenso
+#map2 = affine_map<(d0, d1, d2, d3, d4) -> (d0, d1, d3, d4)>
+#map3 = affine_map<(d0, d1, d2, d3, d4) -> (d2, d3, d4)>
+#map4 = affine_map<(d0, d1, d2, d3, d4) -> (d0, d1, d2)>
+util.func public @quantized_matmul(%arg0: tensor<4096x32x128xi8>, %arg1: tensor<1x1x32x128xf32>) -> tensor<1x1x4096xf32> {
```
I got thrown by this a bit. I guess this doesnt collapse cause of the `linalg.generic` having two `reduction` loops. If it had only one, then the `d0, d1` is collapsible right? Maybe worth adding a test for that.

IanWood1 replied:
Thank you, I added a test for this and found that I wasn't handling mapping between producer <-> consumer loops. I was just bailing in projected cases more than needed.
